### PR TITLE
add umap cuml notebook

### PR DESCRIPTION
**Pull request by quasiben**
cc @cjnolet 

@lmcinnes if you have a minute would you mind looking over this notebook.  I want to make sure attributions are correctly added

### CODE REVIEW

**Comment by lmcinnes** [comment]
Sorry for being a little slow -- I will try to get to this over the weekend. It already looks great at first glance!

**Comment by cjnolet** [comment]
I just realized I was incorrect in a couple of my review comments so I’ve adjusted my review  accordingly.

File: cpu_comparisons/umap-cuml.ipynb [hunk 1]
```diff
@@ -0,0 +1,463 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[UMAP](https://umap-learn.readthedocs.io/en/latest/) is a powerful dimensionality reduction tool which NVIDIA recently port GPUs with a python interface.  In this notebook we will demostrate basic usage, plotting, and timing comparisons between the CUDA (GPU) version of UMAP and the original CPU version of UMAP "
+   ]
+  },
```
Comment (lmcinnes) [inline cpu_comparisons/umap-cuml.ipynb]: Minor typo, probably should be
"... is a powerful dimensionality reduction tool which NVIDIA recently **ported** GPUs with a python interface ..."

File: cpu_comparisons/umap-cuml.ipynb [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    "    'Sneaker',\n",
+    "    'Bag',\n",
+    "    'Ankle boot']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Just as the original author of UMAP Leland McInnes does in the [UMAP docs](https://umap-learn.readthedocs.io/en/latest/supervised.html), we can plot the results and show the separation bewtween the various classes defined above."
```
Comment (lmcinnes) [inline cpu_comparisons/umap-cuml.ipynb]: Typo "bewtween" should be between

File: cpu_comparisons/umap-cuml.ipynb [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
...x5HMKhwbpeAeOGQGcdVV0X/o0NbfXs3/JqQwObGfah032PjzIPscLlG7dh6c6yISPnMSB1B4iGw6Q3h5HzarCeng/6Y19+M4ay/Bd24jcuxPdaVI1YzJDZyVw+TT4QCkRc5jO8Mu48BOYPob6rxbR7S+j99qXMGt9qEye5LI2nAsriD+8j9zOw7lfmtsg8sDuwx90Nkd6eVdhW6kLbXRQJ7Np4G37SDaawFVc9I90LyGEEEK8yySHQRx90bZC1SMMwMYwQDM0PBUQmO7Dtka/rXuqKKyuPiodocfjoj9QDiqPjkKrWgDFEwADTdOomOJD05yF4kjqNSV6vFVguMknINOTQY3GwsOrIwy8MITDZ1La6OHNFPSsQou2YKx6mMlzEvxL3wrGVmRxj6uj/HQfxePcjFnoQUdn4Y3XM/Vr15Pc3UNlfAg94KJ4fhnGkikkDBfOby2mxHSya68L2+tgTpNiklXNOGcJ+Z1Rym9fRPUZdQTveAVXcy8poCubZ9KtjZR/ronyyRPxzq1mIN0Ouyz2L8tz6n03UjJ1DC33/Yn+5csZfvjblHRsY+Cbq0gFDfwfmcC4/72U4Kenkl7fS/KR/bA3zbk/vhXV5CTnyFLvm4oWU+z93hP0XvcS/bevZWj7PqKv+Om+8C/g0kmv6iHXlsCo9lD9q9MJ3jyffGchmKu6ZylmfRGquzBy4D2/AeeM4OGrOJwhsawVAP9HmzBqvX+1i/Su2P53dCghhBBCHE8ywiCOOtX6LOTjoDlQVp5cxMJZerirGQfvpnopRA4mODyQizFmOA66s7DdDIA7CNFWID9aetXmtbU8N9/TS++2JOf8sDD1yHZU4qrpQ0MDdKovqQBMot1JvEED02WAVhhkUEDc4cafS2FZCk90gFSfA9OjAxp7myPUltpsfTRCcJyThgV+mr/ybRKDCkNLMLK8neDNc8jNraDzqQOMmT6Z1Moe2h/ZSdFdZxK9ZydWMEY/W/F3d1PT7iC48FxqL+xhcIyX7JzpPLhO45TJI7j846nrLmLD7tXs+NwDnPnYbfzh81eSGYqy+JcK39Qaxpxewvhzq9n3fIrIo3soD02k7uJF6F4XoQ8sZsV3b6NkTh3WSRNxTSrF0ejnlH+5jq1DzzP44k5K4wH23reXMx76Flt++zCJRYrWp9aSOxCn+menoBs6kZ9sxzmjgq7LnkR
```
[the remark this answers is not in the document]
Comment (lmcinnes) [reply cpu_comparisons/umap-cuml.ipynb]: I agree with @cjnolet here -- trustworthiness (while not a perfect quality measure) is probably a more meaningful score in this context, and is relatively easy to compute given the scikit-learn implementation (it is just potentially quite time consuming for a large dataset like fashion-mnist). It might be reasonable to compute trustworthiness on just the test set (small enough that the sklearn implementation may not choke, since it computes an all-pairs distance matrix).

File: cpu_comparisons/umap-cuml.ipynb [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+    "# display a Nike? sneaker\n",
+    "pixels = train[0].reshape((28, 28))\n",
+    "plt.imshow(pixels, cmap='gray')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "There is cost between moving data between host memory and device memory (GPU memory) and we will include that core when comparing speeds"
```
[the remark this answers is not in the document]
Comment (quasiben) [reply cpu_comparisons/umap-cuml.ipynb]: Ok, I'll remove the `gdf` creation out of the timing measurement

File: cpu_comparisons/umap-cuml.ipynb [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+    "record_data = (('fea%d'%i, data[:,i]) for i in range(data.shape[1]))\n",
+    "gdf = cudf.DataFrame(record_data)\n",
+    "g_embedding = cumlUMAP(n_neighbors=5, init=\"spectral\").fit_transform(gdf)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`gdf` is a GPU backed dataframe -- all the data is store in the device memory of the GPU.  With the data converted, we can apply UMAP with *naive* inputs.  We can improve the `cumlUMAP` constructor by doing the `data->gdf` conversion automatically.  Additionally, it should be noted that within cuml, [FAISS](https://github.com/facebookresearch/faiss) is used for extremely fast kNN and it's limited to single precision.  `cumlUMAP` will automatically downcast to `float32` when needed"
```
[the remark this answers is not in the document]
Comment (quasiben) [reply cpu_comparisons/umap-cuml.ipynb]: Originally I was using default inputs and was using *naive* to mean not much care was put into selecting those particular values.  I'll improve this sentence.  

As for the second note, I'm going to remove this line altogether.  I wrote that line when, I thought, cuml did not provide this feature.  my apologies

File: cpu_comparisons/umap-cuml.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (quasiben) [reply cpu_comparisons/umap-cuml.ipynb]: Thank you @cjnolet @lmcinnes .  @cjnolet suggested I do this prior to me post in the notebook and I neglected to do it.  I'll measure those values tonight/tomorrow morning

[the remark this answers is not in the document]
Comment (quasiben) [reply cpu_comparisons/umap-cuml.ipynb]: And thank you for the detailed explanation as well.  I was not very familiar with this scoring

**Comment by quasiben** [comment]
@cjnolet @lmcinnes I very much appreciate the attentive eye and comments.  I'm making changes now and will post back shortly

**Comment by taureandyernv** [comment]
@quasiben Looking over what you're doing, i think it is great work.  I made a few minor changes to the import cells so that the user doesn't have to figure out which are pip installs and which are conda.  However, i've gotten this error using hdbscan.  I hear that you're removing it from your notebook.  Can you please let me know when you do that?

**Comment by quasiben** [comment]
@taureandyernv done!

**Comment by cjnolet** [comment]
This PR LGTM, pending any updates based on the supervised training feature that's going to be added in 0.7.

File: cpu_comparisons/umap-cuml.ipynb [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+   "source": [
+    "%%timeit\n",
+    "g_embedding = cumlUMAP(n_neighbors=5, init=\"spectral\").fit_transform(gdf)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "`gdf` is a GPU backed dataframe -- all the data is stored in the device memory of the GPU.  With the data converted, we can apply the `cumlUMAP` the same inputs as we do for the standard UMAP.  Additionally, it should be noted that within cuml, [FAISS] https://github.com/facebookresearch/faiss) is used for extremely fast kNN and it's limited to single precision.  `cumlUMAP` will automatically downcast to `float32` when needed."
```
Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "`gdf` is a GPU backed dataframe -- all the data is stored in the device memory of the GPU.  With the data converted, we can pass the same inputs to `cumlUMAP` as we do for the standard UMAP.  Additionally, it should be noted that within cuml, [FAISS] https://github.com/facebookresearch/faiss) is used for extremely fast kNN and it's limited to single precision.  `cumlUMAP` will automatically downcast to `float32` when needed."
```

Comment (pentschev) [reply cpu_comparisons/umap-cuml.ipynb]: The sentence sounds a little odd to me. Just nitpicking, feel free to reject.

File: cpu_comparisons/umap-cuml.ipynb [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Compute Runtimes of UMAP training (lower is better):\n",
+    "- cumlUMAP (GPU): 10.5\n",
+    "- UMAP (CPU): 100 seconds\n",
+    "\n",
+    "cumlMAP is ~9.5x speed up over UMAP CPU.  We should also qualitatively compare output accuracy.  We can do this easily by plotting the embeddings and noting simlarites in the plot"
```
Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "cumlUMAP is ~9.5x speed up over UMAP CPU.  We should also qualitatively compare output accuracy.  We can do this easily by plotting the embeddings and noting simlarites in the plot"
```

File: cpu_comparisons/umap-cuml.ipynb [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+    "    'Sneaker',\n",
+    "    'Bag',\n",
+    "    'Ankle boot']"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Just as the original author of UMAP, Leland McInnes, states in the [UMAP docs](https://umap-learn.readthedocs.io/en/latest/supervised.html), we can plot the results and show the separation between the various classes defined above."
```
Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "Just as stated in the [UMAP docs](https://umap-learn.readthedocs.io/en/latest/supervised.html) by Leland McInnes, the original author of UMAP, we can plot the results and show the separation between the various classes defined above."
```

Comment (pentschev) [reply cpu_comparisons/umap-cuml.ipynb]: Nitpicking once more, feel free to reject. :)

File: cpu_comparisons/umap-cuml.ipynb [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
.../7GLT1Q9iOa+coqdOovSWOVhHFVAZGUf5wlkIcQUlnsJbMMihl95noPUAzcZdpMwpNtxyP2O2LkX9Tj6hg90kXumhqHos0x64jlmP/4Cud7YjR2LoyrJwTK3EUpJDKqIQG4gRqusmmYhiOqUUANuF1WRcOQohU480IpPw3h5UWcX3eD2Gmiwcd0xB8cSIb+0nsaWP7LumYl5YRuYN49NzMfwNxRdOj4CUY0TMMeH71R4kWSLa50U/MgvH7ZNRUqBGFNRAEtFhwu/rwRyJMKbcT+jZRpTWAObyI/3lPM/u12Z+1mg0f5cWMPyP0hfkMfyPv8c6aQKFN1yTTkWyW49ax2QRmfTgdSj7XiCVdwqC/dNnfdZoNJr/RS7vFta88xArTr3h8LKi+ZM48cWfoAzTsaN/BQk5hntnE30f1KKi0vrGBnbf8yz+1CCh8tkYrIWHt7UW2zFbzGSHZqIvGEbpslPJK6pm2MIZHD/nYjKsedTe9jg5OS2MP2MG4WVNDD69h/JL52EozyIWDaKrycR2bimxSAtZYg4jd11B0e/mk/2tCeRNHcX7i37CwNYGVFlBdceI7OohNuBD5zCDQUp3YJ5mx3puBeaSHFIdYUKvt2CclEfRswuwzC9Bb7WQ2NKP/dxqEo1eSIGgT1+OYzv6ia5oQ+4IYxjlSM/hoOfoq7UCRTfMIOuEarDoiPV4kYrNKN4E0W3pB0f6vEx0so6sxU4Es46im8eSmwd6u5F4a4hoTwQhy0DWLROJG6IEvQOfwyeu0Wi+qLSUpP9hgpR+umSfPhXnO68eXh5Y/xFJt5uMk+ahy7DRc+dLBHf8hhGvTUEwGFD9LZBRjih+8umURqPRfN7kRJKUEKfLuw1TUzaWolzkEoGBvkOITjvhhJeY248pNxNRr6Nwbg0xOYzDWIxO1DP6urPIvWIiBwM7EduSxAYCFJsmYtNnH7UfteE5kEysPfttbHePQx4rUWSpIvRRO1ufeYjcX89g/g+cGAnC2NMxlI5g642PIvoVRj92Ea2/ehfPY7WMXnE20fWdxGfosBYWojZFaXpsOa2CGX9jB0oqhaKoiJJAtNfL6DvPJ3RKCvePt2Bs85IbDHDgvHeQbCZMUTPKYJzYum5im3tBVUmRBKsO951bUT0JdCMySDUHQAeht9uRiizYLh6BrsBC8E/N+H9Xlw4cjALE05M2dP9
```
Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "And side-by-side we can see that the separation is very similar between both GPU and CPU versions of UMAP.  Both cluster Ankle Boot, Sneaker, and Sandal together (Green/Yellow/Purple), both separate Trousers and Bags entirely, and both have T-shirts, Shirts, Coats, etc. all mixed together."
```

Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "Additionally, we can also quantitatively compare the performance of `cumlUMAP` (GPU UMAP) to the reference/original implementation (CPU UMAP) using the [trustworthiness score](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/manifold/t_sne.py#L395).  From the docstring:\n",
```

Comment (pentschev) [inline cpu_comparisons/umap-cuml.ipynb]:
```suggestion
    "Like `t-SNE`, UMAP tries to capture both global and local structure and thus, we can apply the `trustworthiness` of the `embedding/g_embedding` data against the original input.  With a higher score we are demonstrating that the algorithm does a better job of local structure retention.  As [Corey Nolet](https://github.com/cjnolet) notes:\n",
```

Comment (pentschev) [reply cpu_comparisons/umap-cuml.ipynb]: Sorry, I just noticed GitHub does a terrible job in visualization here. The suggestion is a typo fix on line 356, not replacing the image.

Comment (pentschev) [reply cpu_comparisons/umap-cuml.ipynb]: Same here, typo on line 395.

Comment (pentschev) [reply cpu_comparisons/umap-cuml.ipynb]: And another type on line 400.

File: cpu_comparisons/umap-cuml.ipynb (hunk 6 above)
Comment (quasiben) [reply cpu_comparisons/umap-cuml.ipynb]: 👍

**Comment by quasiben** [comment]
Thanks @pentschev updated with fixes

**Comment by pentschev** [comment]
Cool, thanks @quasiben for this nice notebook!

**Comment by taureandyernv** [comment]
Merged and resolved in PR #36 .  Thanks again @cjnolet @quasiben @lmcinnes and @pentschev
